src/bloqade/geometry/dialects/grid/types.py

```diff
@@ -377,6 +377,84 @@ def shift(self, x_shift: float, y_shift: float) -> "Grid[NumX, NumY]":
             y_init=self.y_init + y_shift if self.y_init is not None else None,
         )
 
+    def shift_subgrid_x(
+        self, x_indices: ilist.IList[int, Nx] | slice, x_shift: float
+    ) -> "Grid[NumX, NumY]":
+        """Shift a sub grid of grid in the x directions.
+
+        Args:
+            grid (Grid): a grid object
+            x_indices (float): a list/ilist of x indices to shift
+            x_shift (float): shift in the x direction
+        Returns:
+            Grid: a new grid object that has been shifted
+        """
+        indices = get_indices(len(self.x_spacing) + 1, x_indices)
+
+        def shift_x(index):
+            new_spacing = self.x_spacing[index]
+            if index in indices and (index + 1) not in indices:
+                new_spacing -= x_shift
+            elif index not in indices and (index + 1) in indices:
+                new_spacing += x_shift
+            return new_spacing
+
+        new_spacing = tuple(shift_x(i) for i in range(len(self.x_spacing)))
+
+        assert all(
+            x >= 0 for x in new_spacing
+        ), "Invalid shift: column order changes after shift."
```

Comment:
Since this is a user input we should make this an exception instead of an assertion. Generally Assertions are used to enforce internal logic inside a code to catch potential bugs while exceptions are indicating invalid inputs from a user.